// bson/src/double.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Double = void 0;
const bson_value_1 = require("./bson_value");
const utils_1 = require("./parser/utils");
/**
 * A class representation of the BSON Double type.
 * @public
 * @category BSONType
 */
class Double extends bson_value_1.BSONValue {
    get _bsontype() {
        return 'Double';
    }
    /**
     * Create a Double type
     *
     * @param value - the number we want to represent as a double.
     */
    constructor(value) {
        super();
        if (value instanceof Number) {
            value = value.valueOf();
        }
        this.value = +value;
    }
    /**
     * Access the number value.
     *
     * @returns returns the wrapped double number.
     */
    valueOf() {
        return this.value;
    }
    toJSON() {
        return this.value;
    }
    toString(radix) {
        return this.value.toString(radix);
    }
    /** @internal */
    toExtendedJSON(options) {
        if (options && (options.legacy || (options.relaxed && isFinite(this.value)))) {
            return this.value;
        }
        if (Object.is(Math.sign(this.value), -0)) {
            // NOTE: JavaScript has +0 and -0, apparently to model limit calculations. If a user
            // explicitly provided `-0` then we need to ensure the sign makes it into the output
            return { $numberDouble: '-0.0' };
        }
        return {
            $numberDouble: Number.isInteger(this.value) ? this.value.toFixed(1) : this.value.toString()
        };
    }
    /** @internal */
    static fromExtendedJSON(doc, options) {
        const doubleValue = parseFloat(doc.$numberDouble);
        return options && options.relaxed ? doubleValue : new Double(doubleValue);
    }
    inspect(depth, options, inspect) {
        inspect ?? (inspect = utils_1.defaultInspect);
        return `new Double(${inspect(this.value, options)})`;
    }
}
exports.Double = Double;
