// bson/src/int_32.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Int32 = void 0;
const bson_value_1 = require("./bson_value");
const utils_1 = require("./parser/utils");
/**
 * A class representation of a BSON Int32 type.
 * @public
 * @category BSONType
 */
class Int32 extends bson_value_1.BSONValue {
    get _bsontype() {
        return 'Int32';
    }
    /**
     * Create an Int32 type
     *
     * @param value - the number we want to represent as an int32.
     */
    constructor(value) {
        super();
        if (value instanceof Number) {
            value = value.valueOf();
        }
        this.value = +value | 0;
    }
    /**
     * Access the number value.
     *
     * @returns returns the wrapped int32 number.
     */
    valueOf() {
        return this.value;
    }
    toString(radix) {
        return this.value.toString(radix);
    }
    toJSON() {
        return this.value;
    }
    /** @internal */
    toExtendedJSON(options) {
        if (options && (options.relaxed || options.legacy))
            return this.value;
        return { $numberInt: this.value.toString() };
    }
    /** @internal */
    static fromExtendedJSON(doc, options) {
        return options && options.relaxed ? parseInt(doc.$numberInt, 10) : new Int32(doc.$numberInt);
    }
    inspect(depth, options, inspect) {
        inspect ?? (inspect = utils_1.defaultInspect);
        return `new Int32(${inspect(this.value, options)})`;
    }
}
exports.Int32 = Int32;
